pymc_bart/pgbart.py

```diff
@@ -265,47 +255,31 @@ def systematic(self, normalized_weights):
         Note: adapted from https://github.com/nchopin/particles
         """
         lnw = len(normalized_weights)
-        single_uniform = (self.uniform.random() + np.arange(lnw)) / lnw
+        single_uniform = (self.uniform.rvs() + np.arange(lnw)) / lnw
         return inverse_cdf(single_uniform, normalized_weights)
 
     def init_particles(self, tree_id: int) -> np.ndarray:
         """Initialize particles."""
         p0 = self.all_particles[tree_id]
-        p1 = p0.copy()
-        p1.sample_leafs(
-            self.sum_trees,
-            self.m,
-            self.normal,
-            self.shape,
-        )
-        # The old tree and the one with new leafs do not grow so we update the weights only once
-        self.update_weight(p0, old=True)
-        self.update_weight(p1, old=True)
-        particles = [p0, p1]
+        # The old tree does not grow so we update the weight only once
+        self.update_weight(p0)
+        particles = [p0]
 
         for _ in self.indices:
             particles.append(
-                ParticleTree(self.a_tree, self.uniform_kf.random() if self.tune else p0.kfactor)
+                ParticleTree(self.a_tree, self.uniform_kf.rvs() if self.tune else p0.kfactor)
             )
 
-        return np.array(particles)
+        return particles
 
-    def update_weight(self, particle, old=False):
+    def update_weight(self, particle):
         """
         Update the weight of a particle.
-
-        Since the prior is used as the proposal,the weights are updated additively as the ratio of
-        the new and old log-likelihoods.
         """
         new_likelihood = self.likelihood_logp(
             (self.sum_trees_noi + particle.tree._predict()).flatten()
         )
-        if old:
-            particle.log_weight = new_likelihood
-        else:
-            particle.log_weight += new_likelihood - particle.old_likelihood_logp
-
-        particle.old_likelihood_logp = new_likelihood
+        particle.log_weight = new_likelihood
```

Comment:
Why do we not need to `- particle.old_likelihood_logp` anymore?

Reply:
We use the likelihood as the weight. The justification for the new algorithm is mostly empirical and based on improved mixing. The new algorithm is somewhat closer to a Particle Bootstrap Filter. The proposal is described as having 4 terms. 


* node depth: The probability that a node at depth $d = (0,1,2,...)$ is non-terminal is given by $\alpha^d$. This prior was proposed and studied Rockova. 
* splitting values: Uniform over the observed values.
* splitting variable: Short version, we sample from a categorial distribution that we learn during the tuning phase. Long version, we compute the distribution over the splitting variables from the data. We begin with a flat categorical distribution, $\alpha_{\text{vec}}$, i.e., all covariates have the same chance of being used as splitting variables. During the tuning phase, we continuously update $\alpha_{\text{vec}}$ based on counting the splitting variables in the accepted trees. After the tuning phase, we fix this distribution and use it to sample from the splitting variables.
* leaf values: Short version, we use a Gaussian centered at the mean of the current sum of trees, evaluated at the proper partition, i.e. we innovate by adding Gaussian noise to the current solution. Long version, we use $\mathcal{N}(\mu_\text{pred}, {\varepsilon^2})$, where $\mu_\text{pred}$ is computed as the mean of the current sum of trees divided by the number of trees \texttt{m}. $\varepsilon$ is computed from $Y$, being $\varepsilon = \frac{3}{\sqrt{m}}$ for binomial data and  $\varepsilon = \frac{Y_\text{std}}{\sqrt{m}}$ for data other than binomial. Because the prior variance on the leaf node parameters can be model dependent during tuning new particles are created with a scale factor $kf$, that multiplies $\varepsilon$. We find that sampling $kf \sim \mathcal{U}(0.33, 0.75)$ provides good results across different datasets. After tuning, the new particles inherit the $kf$ value from the particle they are going to replace.

Reply:
yes, I think that was described in https://arxiv.org/abs/1502.04622 that we use as the source for our first implementation.

Reply:
We don't have proof the implementation is correct. We only know that empirically it provides reasonable results in terms of fit and mixing. As you said it seems the final resampling is enough to correct that, we have discussed this with a couple of colleagues and they have the same impression.  

https://github.com/JingyuHe/XBART is another BART method, that the authors acknowledge is not strictly correct but still provides good answers. I read their article a while ago and I did not have the time to take another look at that article. If you want we can have a talk to further discuss this.

Reply:
We have done that comparison and we are now running a comparison with a BART method implemented in R. And with other regression methods implemented in PyMC/Bambi, like splines and GPs.